In [1]:
from pathlib import Path
import pandas as pd

project_dir = Path(r"C:\Users\HP-ZBOOK i7\graphrag_test\pilot_01")
csv_path = project_dir / "inspection" / "04_relationships_all.csv"

relationships = pd.read_csv(csv_path)

print("Number of relationships:", len(relationships))
print("Columns:", relationships.columns.tolist())

relationships.head()

Number of relationships: 414
Columns: ['id', 'human_readable_id', 'source', 'target', 'description', 'weight', 'combined_degree', 'text_unit_ids']


,id,human_readable_id,source,target,description,weight,combined_degree,text_unit_ids
0,ee487943-e36d-4146-8e2e-b7a9a1fd3067,0,"BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENE...",ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,"The BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT,...",19.0,48,['d7219f76b31818b7b13bf08dda81003d752e53f76608...
1,e87db637-71f3-4134-8a02-954d1e518207,1,HUBERT FECHNER,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,Hubert Fechner is the principal author of the ...,8.0,41,['d7219f76b31818b7b13bf08dda81003d752e53f76608...
2,316e1382-2ca7-477e-9d9f-c5f6bab0b3ba,2,LEONORE GEWESSLER,"BUNDESMINISTERIUM FÜR KLIMASCHUTZ, UMWELT, ENE...",Leonore Gewessler is the Federal Minister lead...,10.0,11,['d7219f76b31818b7b13bf08dda81003d752e53f76608...
3,86b0f8da-0a86-48c8-990d-8fbbe165249b,3,LEONORE GEWESSLER,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,"Leonore Gewessler represents, promotes, and ov...",9.0,41,['d7219f76b31818b7b13bf08dda81003d752e53f76608...
4,c2727a60-8f2d-4b7f-9fd0-25cd463e97c7,4,ÖSTERREICH,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,ÖSTERREICH (Austria) is both the creator and p...,67.0,121,['d7219f76b31818b7b13bf08dda81003d752e53f76608...


In [2]:
pd.set_option("display.max_colwidth", 120)

relationships[
    [
        "source",
        "target",
        "description",
        "weight",
        "combined_degree",
        "text_unit_ids",
    ]
].sort_values(
    by="combined_degree",
    ascending=False
).head(20)

,source,target,description,weight,combined_degree,text_unit_ids
101,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,ÖSTERREICH,"The ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE, also known as the Austrian Photovoltaic Strategy, is a comprehensive nat...",34.0,121,['06c273ec9919eb4693462fdacebd50536ccdcf47bacc47b2b30c8a69005636b654b4919dc579696a4a7abbde4a00511a66f0c0be040f4ffc90...
4,ÖSTERREICH,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,ÖSTERREICH (Austria) is both the creator and primary beneficiary of the ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE (Aust...,67.0,121,['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...
80,ÖSTERREICH,PV-ANLAGE,"Broad adoption of PV-Anlagen in Austria is a direct result of the national strategy and incentives, positioning Aust...",8.0,102,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...
151,ÖSTERREICH,BMK,BMK is Austria's federal ministry responsible for coordinating renewable energy policies and stakeholder involvement.,8.0,98,['41bd25afdc1cce74b36dd6db5e44e91354172fda16475ab6a3f6cf99bddc720aa91cb56b5ea1b1c4774c8c97017d6bb03c52ba2676d7cfaab1...
22,BMK,ÖSTERREICH,"BMK, or the Federal Ministry for Climate Action, Environment, Energy, Mobility, Innovation and Technology, is Austri...",27.0,98,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...
90,ENERGIEGEMEINSCHAFTEN,ÖSTERREICH,Energy communities are established across Austria to facilitate citizen participation in the energy transition.,8.0,96,['f9e54ba797ff2b80a5f0e5f24e3901236ac4130e4e84d01615e67bc810b290cd5ac7c08f71a782797073050d512eda2695c3659569deba0d4b...
21,KLIMA- UND ENERGIEFONDS,ÖSTERREICH,The Climate and Energy Fund provides crucial financial support for national renewable energy and PV projects.,8.0,96,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...
314,ÖSTERREICH,ENERGIEGEMEINSCHAFTEN,ÖSTERREICH (Austria) has implemented the concept of ENERGIEGEMEINSCHAFTEN (energy communities) as an innovative mode...,15.0,96,['dafd7f2f39e49c87969ec1360e9fae0af9de6f290655205c8c120dd24dcc85da04f0536237f7f9529d31deed892d5abae3b8f58fb66ae9e497...
200,ÖSTERREICH,KLIMA- UND ENERGIEFONDS,The Climate and Energy Fund is a national Austrian organization responsible for funding PV projects and energy stora...,8.0,96,['2d9fdebd364caf6ea226dc0ee6624e0c2d54fc65848a9cd18f9b868675333d5be302c20a6cbb1b4f9503e536f5051a60a8783cd3657ab2cf1f...
26,BUND,ÖSTERREICH,The Bund operates at the federal level within Austria to coordinate and promote national energy policies and the PV ...,8.0,96,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...


In [3]:
top_connected_relationships = relationships.nlargest(
    10, "combined_degree"
)

top_weight_relationships = relationships.nlargest(
    10, "weight"
)

random_relationships = relationships.sample(
    n=25,
    random_state=42
)

relationship_audit_sample = (
    pd.concat(
        [
            top_connected_relationships,
            top_weight_relationships,
            random_relationships,
        ]
    )
    .drop_duplicates(subset="id")
    .reset_index(drop=True)
)

print("Relationship audit sample size:", len(relationship_audit_sample))

relationship_audit_sample[
    [
        "source",
        "target",
        "description",
        "weight",
        "combined_degree",
        "text_unit_ids",
    ]
]

Relationship audit sample size: 42


,source,target,description,weight,combined_degree,text_unit_ids
0,ÖSTERREICH,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,ÖSTERREICH (Austria) is both the creator and primary beneficiary of the ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE (Aust...,67.0,121,['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...
1,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,ÖSTERREICH,"The ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE, also known as the Austrian Photovoltaic Strategy, is a comprehensive nat...",34.0,121,['06c273ec9919eb4693462fdacebd50536ccdcf47bacc47b2b30c8a69005636b654b4919dc579696a4a7abbde4a00511a66f0c0be040f4ffc90...
2,ÖSTERREICH,PV-ANLAGE,"Broad adoption of PV-Anlagen in Austria is a direct result of the national strategy and incentives, positioning Aust...",8.0,102,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...
3,BMK,ÖSTERREICH,"BMK, or the Federal Ministry for Climate Action, Environment, Energy, Mobility, Innovation and Technology, is Austri...",27.0,98,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...
4,ÖSTERREICH,BMK,BMK is Austria's federal ministry responsible for coordinating renewable energy policies and stakeholder involvement.,8.0,98,['41bd25afdc1cce74b36dd6db5e44e91354172fda16475ab6a3f6cf99bddc720aa91cb56b5ea1b1c4774c8c97017d6bb03c52ba2676d7cfaab1...
5,KLIMA- UND ENERGIEFONDS,ÖSTERREICH,The Climate and Energy Fund provides crucial financial support for national renewable energy and PV projects.,8.0,96,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...
6,BUND,ÖSTERREICH,The Bund operates at the federal level within Austria to coordinate and promote national energy policies and the PV ...,8.0,96,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...
7,ENERGIEGEMEINSCHAFTEN,ÖSTERREICH,Energy communities are established across Austria to facilitate citizen participation in the energy transition.,8.0,96,['f9e54ba797ff2b80a5f0e5f24e3901236ac4130e4e84d01615e67bc810b290cd5ac7c08f71a782797073050d512eda2695c3659569deba0d4b...
8,ÖSTERREICH,BUND,"The Bund serves as the federal government of Austria, establishing laws and regulations for the entire country.",10.0,96,['23663988970639bea940978ccbd9150fd292856f2f31a1d587392606a64713d5dfeade66f1beea67d1cc0b0e93012337685e0eca3267f98c16...
9,ÖSTERREICH,KLIMA- UND ENERGIEFONDS,The Climate and Energy Fund is a national Austrian organization responsible for funding PV projects and energy stora...,8.0,96,['2d9fdebd364caf6ea226dc0ee6624e0c2d54fc65848a9cd18f9b868675333d5be302c20a6cbb1b4f9503e536f5051a60a8783cd3657ab2cf1f...


In [4]:
relationship_review = relationship_audit_sample[
    [
        "id",
        "source",
        "target",
        "description",
        "weight",
        "combined_degree",
        "text_unit_ids",
    ]
].copy()

relationship_review["source_supported"] = ""
relationship_review["primary_classification"] = ""
relationship_review["direction_correct"] = ""
relationship_review["endpoints_correct"] = ""
relationship_review["description_problem"] = ""
relationship_review["likely_pipeline_stage"] = ""
relationship_review["audit_notes"] = ""

relationship_review.head()

,id,source,target,description,weight,combined_degree,text_unit_ids,source_supported,primary_classification,direction_correct,endpoints_correct,description_problem,likely_pipeline_stage,audit_notes
0,c2727a60-8f2d-4b7f-9fd0-25cd463e97c7,ÖSTERREICH,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,ÖSTERREICH (Austria) is both the creator and primary beneficiary of the ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE (Aust...,67.0,121,['d7219f76b31818b7b13bf08dda81003d752e53f7660831d5e201d105d2bc1fc19e8cc49ca6cd21e08570939d7fb3e87b45f0f5feb8575df195...,,,,,,,
1,0369af84-01bf-4ff5-bf9f-edd6fa8c1fdb,ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE,ÖSTERREICH,"The ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE, also known as the Austrian Photovoltaic Strategy, is a comprehensive nat...",34.0,121,['06c273ec9919eb4693462fdacebd50536ccdcf47bacc47b2b30c8a69005636b654b4919dc579696a4a7abbde4a00511a66f0c0be040f4ffc90...,,,,,,,
2,e19bcd0c-85bf-45d4-b522-9dfe54a4e2d5,ÖSTERREICH,PV-ANLAGE,"Broad adoption of PV-Anlagen in Austria is a direct result of the national strategy and incentives, positioning Aust...",8.0,102,['e498566a78d36b3ce13304157993b9b4269045840f1d63663ab07328cc5c6a9ebc15f47c9ae340cbebf706bc8faebd35ccc4f11ae0aa4cb9a2...,,,,,,,
3,1135fa09-924a-4fd4-ad10-3895fb2dee28,BMK,ÖSTERREICH,"BMK, or the Federal Ministry for Climate Action, Environment, Energy, Mobility, Innovation and Technology, is Austri...",27.0,98,['50fb8f6506c2271985ee7ffe42523d2eb06abbd85be8ea5021ae2d36d260f6f63d177e59ca2da764eeaeb4a953c841282dbf214859a753f968...,,,,,,,
4,a29f9ef9-08a7-4547-9b3f-3bfd1be6cc32,ÖSTERREICH,BMK,BMK is Austria's federal ministry responsible for coordinating renewable energy policies and stakeholder involvement.,8.0,98,['41bd25afdc1cce74b36dd6db5e44e91354172fda16475ab6a3f6cf99bddc720aa91cb56b5ea1b1c4774c8c97017d6bb03c52ba2676d7cfaab1...,,,,,,,


In [5]:
def add_relationship_review(
    row_number,
    source_supported,
    classification,
    direction_correct,
    endpoints_correct,
    description_problem="",
    pipeline_stage="",
    notes="",
):
    mask = relationship_review.index == row_number

    relationship_review.loc[mask, "source_supported"] = source_supported
    relationship_review.loc[mask, "primary_classification"] = classification
    relationship_review.loc[mask, "direction_correct"] = direction_correct
    relationship_review.loc[mask, "endpoints_correct"] = endpoints_correct
    relationship_review.loc[mask, "description_problem"] = description_problem
    relationship_review.loc[mask, "likely_pipeline_stage"] = pipeline_stage
    relationship_review.loc[mask, "audit_notes"] = notes

In [6]:
add_relationship_review(
    row_number=0,
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="Austria is described as creator, primary beneficiary, executor, and continuous implementer of the strategy.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="The strategy concerns Austria, but the description assigns several responsibilities more specifically associated with government institutions."
)

add_relationship_review(
    row_number=1,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Description combines several themes from across the complete strategy.",
    pipeline_stage="DESCRIPTION_SUMMARIZATION",
    notes="The document is explicitly an Austrian national photovoltaic strategy."
)

add_relationship_review(
    row_number=2,
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="Claims that broad PV adoption is a direct result of national strategy and incentives.",
    pipeline_stage="CAUSAL_GENERALIZATION",
    notes="The source discusses incentives and expansion, but does not establish this broad direct causal claim."
)

add_relationship_review(
    row_number=3,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Expands BMK's responsibilities into comprehensive oversight, financing, and delivery of national transformation goals.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="BMK is an Austrian ministry and publisher, but the complete description is broader than one explicit source statement."
)

add_relationship_review(
    row_number=4,
    source_supported="YES",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Relationship is broadly correct but uses a country-to-ministry association rather than a controlled predicate.",
    pipeline_stage="UNCONTROLLED_RELATIONSHIP_SCHEMA",
    notes="BMK is Austria's federal ministry, but the relationship duplicates the reverse BMK-to-Austria connection."
)

add_relationship_review(
    row_number=5,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="The adjective 'crucial' and the broad national funding role are interpretive.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="The fund supports relevant projects, but the relationship description is broader than a direct source statement."
)

add_relationship_review(
    row_number=6,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Relationship is broad and analytically weak.",
    pipeline_stage="RELATIONSHIP_SPECIFICITY",
    notes="The federal level operates within Austria and participates in national energy-policy coordination."
)

add_relationship_review(
    row_number=7,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Description generalizes energy communities across Austria.",
    pipeline_stage="MINOR_DESCRIPTION_GENERALIZATION",
    notes="The source explicitly discusses energy communities as mechanisms for citizen participation in Austria."
)

add_relationship_review(
    row_number=8,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Uses general constitutional knowledge about the Austrian federal government.",
    pipeline_stage="GENERAL_KNOWLEDGE_ADDITION",
    notes="Broadly true, but it adds little solar-market-specific information and duplicates the reverse relationship."
)

add_relationship_review(
    row_number=9,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Summarizes the fund as nationally responsible for PV and storage funding.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="The fund is relevant to Austrian project financing, but the wording implies a broader formal responsibility."
)

In [7]:
add_relationship_review(
    row_number=10,
    source_supported="YES",
    classification="OVERGENERALIZED",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Claims that every federal state is mandated to establish and execute ambitious PV targets.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_GENERALIZATION",
    notes="The source discusses federal-state responsibilities, but the description expands them into a uniform mandatory role."
)

add_relationship_review(
    row_number=11,
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="The ministry is described as owner, principal driver, formulator, and active overseer of implementation.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="The source explicitly identifies the ministry as publisher, but the full operational role is broader than the publication statement."
)

add_relationship_review(
    row_number=12,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="The relationship is explicit, although the description adds broad statements about all national efforts.",
    pipeline_stage="MINOR_DESCRIPTION_GENERALIZATION",
    notes="The source explicitly states Austria's objective of climate neutrality by 2040."
)

add_relationship_review(
    row_number=13,
    source_supported="YES",
    classification="OVERGENERALIZED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="The energy transition is described as a single strategic initiative and Austria as a climate leader.",
    pipeline_stage="ENTITY_AND_RELATIONSHIP_OVERGENERALIZATION",
    notes="The source describes an energy-transition process, but the relationship turns it into a bounded initiative with additional evaluative claims."
)

add_relationship_review(
    row_number=14,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Claims that E-Control oversees reforms and guides Austria's renewable transition, exceeding the specific source wording.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="E-Control is an Austrian regulator and relevant actor, but the description expands its role."
)

add_relationship_review(
    row_number=15,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Adds general knowledge about Austria's EU membership and legal obligations.",
    pipeline_stage="GENERAL_KNOWLEDGE_ADDITION",
    notes="The document discusses EU directives affecting Austria, but the full membership-and-compliance explanation is synthesized."
)

add_relationship_review(
    row_number=16,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Description uses broad language such as primary beneficiaries and main actors.",
    pipeline_stage="MINOR_DESCRIPTION_GENERALIZATION",
    notes="The source explicitly presents citizens as participants in energy communities."
)

add_relationship_review(
    row_number=17,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    pipeline_stage="NONE",
    notes="The source explicitly identifies the Grid Connection Action Plan as an E-Control action plan and cites the July 2023 document."
)

add_relationship_review(
    row_number=18,
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="A bibliographic co-author of one study is described as generally providing research and advice on Austria's markets.",
    pipeline_stage="REFERENCE_MISINTERPRETATION",
    notes="FFE appears in a study citation; the broader advisory relationship to Austria is not directly established."
)

add_relationship_review(
    row_number=19,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="A shared bibliographic citation is converted into a general project-collaboration relationship.",
    pipeline_stage="REFERENCE_MISINTERPRETATION",
    notes="PV Austria and EPOOL are listed together as study contributors, which supports collaboration only indirectly."
)

In [8]:
add_relationship_review(
    row_number=20,
    source_supported="NO",
    classification="UNSUPPORTED",
    direction_correct="NO",
    endpoints_correct="YES",
    description_problem="Two directives listed in the document are presented as directly linked parts of one integrated reform.",
    pipeline_stage="REFERENCE_PROXIMITY_ERROR",
    notes="The source does not explicitly establish a direct relationship between these two directives."
)

add_relationship_review(
    row_number=21,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    pipeline_stage="NONE",
    notes="The bibliography explicitly lists PV Austria studies and guidance produced on behalf of the Climate and Energy Fund."
)

add_relationship_review(
    row_number=22,
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Calls biodiversity solar parks a leading and unique PV type and adds detailed ecological claims.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="The source connects biodiversity solar parks with PV and biodiversity, but the description is more elaborate and evaluative."
)

add_relationship_review(
    row_number=23,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="Two separate initiatives are connected because they pursue similar European PV-production objectives.",
    pipeline_stage="SHARED_THEME_INFERENCE",
    notes="The source discusses both initiatives together, but does not state a direct relationship between them."
)

add_relationship_review(
    row_number=24,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Generalizes that building renovations often trigger PV requirements.",
    pipeline_stage="RELATIONSHIP_GENERALIZATION",
    notes="The source connects renovated buildings with PV integration, but the frequency and legal-trigger wording are synthesized."
)

add_relationship_review(
    row_number=25,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    pipeline_stage="NONE",
    notes="The source explicitly discusses the need for PV installations to connect to and interact with the public electricity grid."
)

add_relationship_review(
    row_number=26,
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="Expands the solar industry's role into manufacturing, implementation support, innovation, and global-market activity.",
    pipeline_stage="RELATIONSHIP_DESCRIPTION_SUMMARIZATION",
    notes="The source connects PV systems and domestic solar industry, but the description combines several broad claims."
)

add_relationship_review(
    row_number=27,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    pipeline_stage="NONE",
    notes="The source explicitly states that the Bund-Länder Dialogue was created jointly and that the federal government participates."
)

add_relationship_review(
    row_number=28,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Claims that climate model regions rely on federal-state policy support.",
    pipeline_stage="RELATIONSHIP_INFERENCE",
    notes="The source discusses established climate model regions in Austria but does not explicitly define this dependency on the Bundesländer."
)

add_relationship_review(
    row_number=29,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="The wording 'can now' may imply a precise temporal or legal status.",
    pipeline_stage="MINOR_DESCRIPTION_GENERALIZATION",
    notes="The source states that active-customer development can enable energy sharing without establishing a separate legal entity."
)

In [9]:
add_relationship_review(
    row_number=30,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="The description combines regulatory oversight and wider cooperation into one relationship.",
    pipeline_stage="MINOR_DESCRIPTION_SUMMARIZATION",
    notes="The source explicitly identifies E-Control and grid operators as actors whose cooperation is important."
)

add_relationship_review(
    row_number=31,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="The phrase 'key demonstration zones' is a summarizing interpretation.",
    pipeline_stage="MINOR_DESCRIPTION_GENERALIZATION",
    notes="The strategy explicitly calls for established climate model regions to become model regions for ambitious PV generation."
)

add_relationship_review(
    row_number=32,
    source_supported="NO",
    classification="UNSUPPORTED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="A university named as co-author of a cited study is described as participating in research funded by the Climate and Energy Fund.",
    pipeline_stage="REFERENCE_MISINTERPRETATION",
    notes="The source citation does not establish that this particular University of Klagenfurt study was funded by the fund."
)

add_relationship_review(
    row_number=33,
    source_supported="PARTIAL",
    classification="OVERGENERALIZED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="Claims that standardization organizations act through BMK to create PV technical standards.",
    pipeline_stage="CO_OCCURRENCE_GENERALIZATION",
    notes="The source lists BMK and standardization organizations among relevant actors but does not define this specific operational relationship."
)

add_relationship_review(
    row_number=34,
    source_supported="PARTIAL",
    classification="REASONABLE_INFERENCE_NOT_EXPLICIT",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="Converts general calls for stakeholder cooperation into a specific joint support relationship.",
    pipeline_stage="CO_OCCURRENCE_INFERENCE",
    notes="Business and industry associations are relevant actors, but the exact joint relationship is synthesized."
)

add_relationship_review(
    row_number=35,
    source_supported="PARTIAL",
    classification="WRONG_DIRECTION_OR_ENDPOINTS",
    direction_correct="QUESTIONABLE",
    endpoints_correct="NO",
    description_problem="The implementation relationship belongs to the new ElWG, not the replaced ElWOG.",
    pipeline_stage="ENTITY_CONFUSION_AND_WRONG_ENDPOINT",
    notes="The source states that ElWG replaces ElWOG and implements EU directives; GraphRAG connected the old ElWOG to the directive."
)

add_relationship_review(
    row_number=36,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="PARTIAL",
    description_problem="The target uses the inconsistent abbreviation NIP rather than the preferred Austrian abbreviation ÖNIP.",
    pipeline_stage="ENTITY_RESOLUTION",
    notes="The strategy explicitly states that its values and scenarios use potential estimates from the integrated Austrian network infrastructure plan."
)

add_relationship_review(
    row_number=37,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    description_problem="The label represents a broad stakeholder group rather than a named entity.",
    pipeline_stage="ENTITY_SCHEMA_LIMITATION",
    notes="The source discusses incentives and opportunities for private individuals to install or participate in photovoltaics."
)

add_relationship_review(
    row_number=38,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    pipeline_stage="NONE",
    notes="The complete strategy is explicitly focused on photovoltaic expansion and its role in Austria's energy transition."
)

add_relationship_review(
    row_number=39,
    source_supported="NO",
    classification="UNSUPPORTED",
    direction_correct="QUESTIONABLE",
    endpoints_correct="YES",
    description_problem="A cited industry association is described as shaping and reporting on the national strategy.",
    pipeline_stage="REFERENCE_MISINTERPRETATION",
    notes="The document cites PV Austria material but does not explicitly establish that the association shaped the Austrian PV Strategy."
)

add_relationship_review(
    row_number=40,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    pipeline_stage="NONE",
    notes="The source explicitly states that nature conservation and spatial planning are primarily within federal-state competence."
)

add_relationship_review(
    row_number=41,
    source_supported="YES",
    classification="DIRECTLY_SUPPORTED_BY_SOURCE",
    direction_correct="YES",
    endpoints_correct="YES",
    pipeline_stage="NONE",
    notes="The source explicitly identifies Mobilitätswende as one of the strategic focus areas in the national FTI Pact."
)

In [11]:
reviewed_relationships = relationship_review[
    relationship_review["primary_classification"] != ""
]

print("Reviewed relationships:", len(reviewed_relationships))
print()
print("Primary classifications:")
print(reviewed_relationships["primary_classification"].value_counts())
print()
print("Source support:")
print(reviewed_relationships["source_supported"].value_counts())

Reviewed relationships: 42

Primary classifications:
primary_classification
DIRECTLY_SUPPORTED_BY_SOURCE         17
REASONABLE_INFERENCE_NOT_EXPLICIT    12
OVERGENERALIZED                       9
UNSUPPORTED                           3
WRONG_DIRECTION_OR_ENDPOINTS          1
Name: count, dtype: int64

Source support:
source_supported
YES        20
PARTIAL    19
NO          3
Name: count, dtype: int64


## Stage 4 — Relationship-Extraction Quality Audit

### Objective

The purpose of this stage was to determine whether relationships extracted by GraphRAG are supported by the Austrian Photovoltaic Strategy.

The original PDF and the associated text units were treated as authoritative evidence. GraphRAG’s relationship descriptions and weights were treated as model-generated outputs requiring independent verification.

The audit examined:

* Whether the source supports the relationship
* Whether the relationship is explicit or inferred
* Whether the description generalizes beyond the source
* Whether the source and target entities are appropriate
* Whether the direction is meaningful
* Whether references or co-occurrence were misinterpreted as substantive relationships

---

## Complete relationship output

GraphRAG extracted **414 relationships** from one source document.

Each relationship contains:

| Field             | Meaning                                         |
| ----------------- | ----------------------------------------------- |
| `source`          | First entity selected by GraphRAG               |
| `target`          | Second entity selected by GraphRAG              |
| `description`     | LLM-generated explanation of the connection     |
| `weight`          | Aggregated GraphRAG relationship-strength value |
| `combined_degree` | Combined graph connectivity of the endpoints    |
| `text_unit_ids`   | Text units associated with the relationship     |

Neither `weight` nor `combined_degree` represents factual confidence.

A high value does not demonstrate that a relationship is true. It may reflect repeated extraction, merged descriptions, frequent entities, overlap, or graph connectivity.

---

## Sampling method

A reproducible sample was created from three components:

* Ten relationships with the highest `combined_degree`
* Ten relationships with the highest GraphRAG `weight`
* Twenty-five randomly sampled relationships
* Removal of duplicated rows appearing in more than one component

A fixed random seed was used:

```python
random_state=42
```

After removing overlaps, the final audit sample contained **42 unique relationships**.

The sample includes:

* Structurally influential relationships
* Relationships given high weight by GraphRAG
* Ordinary randomly selected relationships

This design supports bottleneck diagnosis but is not a simple random sample of all 414 relationships.

---

## Classification results

| Primary classification              |  Count | Approximate share |
| ----------------------------------- | -----: | ----------------: |
| `DIRECTLY_SUPPORTED_BY_SOURCE`      |     17 |             40.5% |
| `REASONABLE_INFERENCE_NOT_EXPLICIT` |     12 |             28.6% |
| `OVERGENERALIZED`                   |      9 |             21.4% |
| `UNSUPPORTED`                       |      3 |              7.1% |
| `WRONG_DIRECTION_OR_ENDPOINTS`      |      1 |              2.4% |
| **Total**                           | **42** |          **100%** |

Only **17 of 42 relationships**, approximately **40.5%**, were classified as directly supported by the source.

The remaining **25 relationships**, approximately **59.5%**, required qualification because they were inferred, overgeneralized, unsupported, or connected to an incorrect endpoint.

---

## Source-support results

| Source support |  Count | Approximate share |
| -------------- | -----: | ----------------: |
| `YES`          |     20 |             47.6% |
| `PARTIAL`      |     19 |             45.2% |
| `NO`           |      3 |              7.1% |
| **Total**      | **42** |          **100%** |

A `PARTIAL` classification means that the source mentions the entities or a narrower connection, but the complete relationship description is not explicitly supported.

The large number of partially supported relationships demonstrates that GraphRAG often begins with genuine source material and then expands it through plausible-sounding interpretation.

---

## Directly supported relationships

Several relationships were explicitly grounded in the source.

### Austria and climate neutrality

```text
ÖSTERREICH → KLIMANEUTRALITÄT 2040
```

The foreword explicitly states the objective of a climate-neutral Austria by 2040.

The basic relationship is directly supported, although the generated description adds broader statements about all national policies and initiatives being aligned with this target.

### Citizens and energy communities

```text
BÜRGER:INNEN → ENERGIEGEMEINSCHAFTEN
```

The source explicitly describes citizens as participants in energy communities and the broader energy-transition process.

### PV installations and the electricity grid

```text
STROMNETZ → PHOTOVOLTAIKANLAGEN
```

The strategy directly discusses the need to connect PV installations to the public electricity grid and identifies grid integration as a major technical challenge.

### PV Strategy and photovoltaics

```text
PHOTOVOLTAIK (PV) → ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE
```

The strategy is explicitly centred on the development and role of photovoltaics in Austria.

### FTI Pact and mobility transition

```text
NATIONALER FTI-PAKT → MOBILITÄTSWENDE
```

The source explicitly identifies `Mobilitätswende` as one of the research focus areas within the national FTI Pact.

These examples show that GraphRAG can identify meaningful relationships when the source states the connection clearly.

---

## Reasonable but non-explicit inferences

Twelve relationships were considered plausible but not directly stated in the exact form generated by GraphRAG.

### BMK and Austria

```text
BMK → ÖSTERREICH
```

BMK is an Austrian federal ministry, and the source identifies it as the publisher. However, the generated description expands the relationship into comprehensive oversight, financing, and delivery of Austria’s transition goals.

The relationship is broadly reasonable, but the complete description is not a direct source statement.

### European Union and Austria

```text
EUROPÄISCHE UNION → ÖSTERREICH
```

The document discusses EU directives and their influence on Austrian energy policy. GraphRAG adds general knowledge about EU membership and compliance obligations.

The resulting description is plausible but partly synthesized.

### Climate model regions and federal states

```text
KLIMAMODELLREGIONEN → BUNDESLÄNDER
```

The source discusses established climate model regions in Austria. GraphRAG infers that they are established within federal states and depend on their policy support.

This may be reasonable, but the dependency is not explicitly established.

### PV Austria and EPOOL

```text
BUNDESVERBAND PV AUSTRIA → EPOOL
```

The two organizations appear together in citations for studies and guidance. GraphRAG interprets this as a general collaboration relationship.

A shared publication supports some connection, but the description is broader than the bibliographic evidence.

These relationships should be retained only if the KG distinguishes explicit facts from inferred associations.

---

## Overgeneralized relationships

Nine sampled relationships were classified as overgeneralized.

### Austria and PV deployment

```text
ÖSTERREICH → PV-ANLAGE
```

GraphRAG states that broad PV adoption in Austria is a direct result of national strategy and incentives.

The source describes policy support and PV expansion but does not establish a single direct causal relationship accounting for broad adoption.

### Austria and the federal states

```text
ÖSTERREICH → BUNDESLÄNDER
```

The source discusses the important responsibilities of federal states, including spatial planning and implementation.

GraphRAG expands this into a claim that each state is mandated to establish and execute ambitious expansion targets. This is broader than the evidence reviewed.

### Ministry and Austrian PV Strategy

```text
MINISTRY → AUSTRIAN PV STRATEGY
```

The source explicitly identifies the ministry as publisher.

GraphRAG additionally describes the ministry as:

* Owner
* Principal driver
* Formulator
* Active implementation supervisor
* Coordinator of regulatory and funding mechanisms

Some of these roles may be plausible, but the complete description exceeds the explicit publication evidence.

### FFE and Austria

```text
FORSCHUNGSSTELLE FÜR ENERGIEWIRTSCHAFT → ÖSTERREICH
```

FFE appears as an author or contributor to a cited study.

GraphRAG converts this bibliographic presence into a general relationship in which FFE provides research and advice on Austrian electricity and flexibility markets.

This is a reference-interpretation error.

### PV systems and the solar industry

```text
PV-ANLAGE → SOLARINDUSTRIE
```

The source connects the PV industry with domestic manufacturing and value creation.

The generated description combines manufacturing, implementation, innovation, domestic markets, and international markets into one broad relationship.

These cases demonstrate that GraphRAG frequently transforms narrower source statements into more comprehensive narratives.

---

## Unsupported relationships

Three sampled relationships were not supported by the source evidence reviewed.

### Relationship between two directives

```text
RICHTLINIE (EU) 2012/27/EU
    →
STROMBINNENMARKTRICHTLINIE 2019/944
```

The relationship description claims that the directives are directly linked as part of an integrated legislative reform.

The document references legal instruments but does not explicitly establish this direct relationship.

The likely cause is proximity within legal or bibliographic text.

### University of Klagenfurt and Climate and Energy Fund

```text
UNI KLAGENFURT → KLIMA- UND ENERGIEFONDS
```

The University of Klagenfurt appears as a contributor to a cited study.

GraphRAG claims that the university participates in research funded by the Climate and Energy Fund. The reviewed citation does not establish that funding relationship.

### PV Austria and the Austrian PV Strategy

```text
BUNDESVERBAND PV AUSTRIA
    →
ÖSTERREICHISCHE PHOTOVOLTAIK-STRATEGIE
```

GraphRAG states that PV Austria helped shape and report on the Austrian PV Strategy.

The source cites PV Austria publications but does not explicitly state that the association shaped the strategy.

These examples show that references and co-occurrence can produce unsupported graph edges.

---

## Wrong endpoint

One sampled relationship contained a material endpoint error:

```text
ELWOG → ERNEUERBAREN-RICHTLINIE 2018/2001
```

The source explains that the new `ElWG` replaces the previous `ElWOG` and serves, among other things, to implement EU directives.

GraphRAG connected the old `ElWOG` to the implementation relationship instead of the new `ElWG`.

This is not merely a broad inference. It changes which law performs the stated function.

The likely cause is entity-name similarity combined with multiple related laws appearing in the same passage.

---

## Direction and relationship-schema limitations

GraphRAG’s raw relationships do not use controlled predicates such as:

```text
IMPLEMENTS
PUBLISHES
HAS_TARGET
SUPPORTS
CONSTRAINS
APPLIES_IN
FUNDED_BY
DERIVED_FROM
```

Instead, they contain a generic source, target, and free-text description.

This creates several problems:

1. The meaning of direction is sometimes unclear.
2. Reverse duplicates can coexist.
3. The same endpoints can carry different narrative descriptions.
4. Automated validation becomes difficult.
5. Relationships cannot be reliably queried by semantic type.

For example, both of the following appear:

```text
AUSTRIA → AUSTRIAN PV STRATEGY
```

```text
AUSTRIAN PV STRATEGY → AUSTRIA
```

Without a controlled predicate, it is unclear whether these mean:

* Austria has the strategy
* The strategy applies in Austria
* Austria created the strategy
* Austria benefits from the strategy
* The strategy concerns Austria

A controlled KG should replace these vague links with explicit predicates.

---

## Weight interpretation

Some relationships have high GraphRAG weights:

```text
AUSTRIA → AUSTRIAN PV STRATEGY
weight = 67
```

```text
AUSTRIAN PV STRATEGY → AUSTRIA
weight = 34
```

These values must not be interpreted as:

* Confidence percentages
* Truth probabilities
* Human validation scores
* Factual reliability measures

In the reviewed sample, high-weight relationships could still contain overgeneralized descriptions.

GraphRAG weight should therefore be treated as an internal structural or extraction-strength value, not as evidence.

---

## Main relationship-error mechanisms

### Causal generalization

The source says that policies support PV expansion. GraphRAG states that adoption is a direct result of those policies.

### Co-occurrence inference

Two actors appear in the same section. GraphRAG creates a specific operational relationship between them.

### Reference misinterpretation

Two organizations or authors appear together in a citation. GraphRAG creates a general collaboration, funding, or advisory relationship.

### Description expansion

A narrow relationship is converted into a long narrative containing roles, goals, effects, and causal claims.

### Entity confusion

Similar policy or law names cause GraphRAG to connect the relationship to the wrong entity, as in the `ElWG` versus `ElWOG` example.

### General-knowledge addition

The description includes correct background knowledge that is not explicitly provided by the supporting source passage.

---

## Representative pipeline traces

| Source evidence                               | GraphRAG output                                   | Problem                                   | Likely pipeline stage     | Recommended fix                             |
| --------------------------------------------- | ------------------------------------------------- | ----------------------------------------- | ------------------------- | ------------------------------------------- |
| Austria targets climate neutrality by 2040    | Austria → Climate Neutrality 2040                 | Directly supported; description broadened | Description summarization | Retain edge, shorten description            |
| Citizens participate in energy communities    | Citizens → Energy Communities                     | Directly supported                        | Extraction successful     | Map to controlled `PARTICIPATES_IN`         |
| PV requires grid connection                   | Grid → PV installations                           | Directly supported                        | Extraction successful     | Normalize direction and predicate           |
| Ministry is publisher                         | Ministry → Strategy as owner and principal driver | Overgeneralized                           | Description generation    | Separate `PUBLISHED_BY` from inferred roles |
| Two organizations occur in one citation       | General collaboration relationship                | Inference from bibliography               | Reference processing      | Mark citation contributors separately       |
| Two directives appear near each other         | Direct legislative relationship                   | Unsupported                               | Proximity/co-occurrence   | Require explicit relational evidence        |
| ElWG replaces ElWOG and implements directives | ElWOG linked to directive                         | Wrong endpoint                            | Entity confusion          | Canonical legal-entity resolution           |
| Vienna appears as publication location        | Vienna → Austria capital relationship             | General knowledge addition                | Description generation    | Restrict to source-grounded claims          |

---

## Severity assessment

**Direct relationship extraction:** Moderate quality
GraphRAG successfully extracts some explicit relationships.

**Inference control:** Poor to moderate quality
Many plausible relationships are not explicitly stated.

**Description precision:** Poor quality
Descriptions frequently expand beyond the narrow source evidence.

**Endpoint resolution:** Moderate concern
The `ElWG`/`ElWOG` example shows that similar entities can cause materially incorrect links.

**Reference handling:** Poor quality
Bibliographic co-occurrence can become unsupported domain relationships.

**Relationship typing:** Major limitation
Free-text relationships lack controlled semantic predicates.

---

## Methodological limitations

The 42-relationship sample is not a simple random sample of all 414 relationships.

It intentionally includes:

* Highly connected relationships
* High-weight relationships
* Random relationships

The reported proportions should therefore be used for pipeline diagnosis rather than treated as exact population-wide error rates.

The classification evaluates one primary problem per relationship. Some relationships have multiple simultaneous problems.

For example, a relationship may be:

* Partially supported
* Overgeneralized
* Directionally ambiguous
* Connected to an alias rather than a canonical entity

Only one primary classification was used for counting, while additional issues were recorded separately.

---

## Stage 4 conclusion

GraphRAG extracted several useful relationships from the Austrian Photovoltaic Strategy, particularly when the source stated the connection explicitly.

However, only 17 of 42 sampled relationships were classified as directly supported. The majority required qualification because GraphRAG introduced inference, broad causal language, generalized descriptions, reference-based associations, or an endpoint error.

The relationship descriptions are substantially less reliable than simple evidence-linked triples. GraphRAG often converts a narrow source connection into a broad narrative containing additional roles, goals, and effects.

For the controlled KG, raw GraphRAG relationships should not be accepted automatically. They should be transformed into:

```text
Canonical source entity
    →
Controlled predicate
    →
Canonical target entity
    →
Source document, page, and chunk
```

Explicit source relationships should be separated from inferred relationships. Inferred edges should carry a distinct status and must not be treated as equivalent to source facts.

The evidence indicates that relationship extraction is a significant bottleneck, particularly in relation to description generalization, reference interpretation, and lack of controlled predicates.
